File: Linked_List/Linked_list.cpp
#include<bits/stdc++.h>
using namespace std;


struct Node{
    int data;
    struct Node* next;
    Node(int data){
        this->data = data;
        next = NULL;
    }
};

struct LinkedList{
    Node* head;
    LinkedList(){
        head = NULL;
    }
    void printList(){
        struct Node* temp = head;
        while(temp != NULL){
            cout << temp->data << " ";
            temp = temp->next;
        }
    }
    void push(int data){
        Node* temp = new Node(data);
        temp->next = head;
        head = temp;
    }
    void push_back(int data){
        Node* temp = new Node(data);

        if (head == NULL) head = temp;
        else{
            Node* last_node = head;
            while(last_node->next != NULL){
                last_node = last_node->next;
            } 
            last_node->next = temp;
        }
    }

    void del_front(){
        // delete linked list from beginning
        Node* temp = head;
        head = head->next;
        delete temp;
    }

    void del_pop(){
        Node* temp = head;
        while(temp->next->next !=NULL){
            temp = temp->next;
        }
        temp->next = NULL;
    }

    void del_position(int position){
        Node* temp = head;
        Node* prev = head;
        for (int i = 0; i < position; i++){
            if (i == 0 && position == 1){
                head = head->next;
                delete(temp);
            }
            else{
                // 이전 포지션 next값을 연결 후 삭제 
                if (i == position-1 && temp){
                    prev->next = temp->next;
                    delete(temp);
                }
                else {
                    // temp -> next 이동 && prev 이동 
                    prev = temp;
                    if (prev == NULL) break; // Position > len(Linked List)
                    temp = temp->next;
                }
            }
        }
    }
};

int main(){
    LinkedList LL;
    LL.push_back(1);
    LL.push_back(2);
    LL.push_back(3);
    LL.push_back(4);
    LL.push_back(5);
    LL.push(0);
    LL.printList();
    cout << "\n";
    LL.del_position(5);
    LL.printList();
}

File: Linked_List/Linked_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None 
    
    def printList(self):
        temp = self.head 
        while(temp):
            print(temp.data, end = " ")
            temp = temp.next 
    
    def push(self, data):
        temp = Node(data)
        temp.next = self.head 
        self.head = temp 
    
    def push_back(self, data):
        temp = Node(data)
        if self.head == None:
            self.head = temp
        else:
            last_node = self.head
            while(last_node.next != None):
                last_node = last_node.next 
            last_node.next = temp 
    
    def del_front(self):
        self.head = self.head.next
        
    def del_pop(self):
        temp = self.head 
        while(temp.next and temp.next.next != None):
            temp = temp.next
        temp.next = None 
        
    def del_position(self, pos):
        temp = self.head
        prev = self.head 
        for i in range(0, pos):
            if i == 0 and pos == 1:
                self.head = self.head.next
            else:
                if i == (pos-1) and temp is not None:
                    prev.next = temp.next 
                else:
                    prev = temp 
                    # position was greater than num of list
                    if prev is None:
                        break
                    temp = temp.next 
            
    
LL = LinkedList()
LL.push_back(0)
LL.push_back(1)
LL.push_back(2)
LL.push_back(3)
LL.push_back(4)
LL.push_back(5)
LL.printList();
print("\n")
LL.del_position(2)
LL.printList();




    
